pkg/receive/writer.go

```diff
@@ -105,13 +115,19 @@ func (r *Writer) Write(ctx context.Context, tenantID string, wreq *prompb.WriteR
 		if ref == 0 {
 			// If not, copy labels, as TSDB will hold those strings long term. Given no
 			// copy unmarshal we don't want to keep memory for whole protobuf, only for labels.
-			labelpb.ReAllocZLabelsStrings(&t.Labels, r.intern)
+			labelpb.ReAllocZLabelsStrings(&t.Labels, r.opts.Intern)
 			lset = labelpb.ZLabelsToPromLabels(t.Labels)
 		}
 
 		// Append as many valid samples as possible, but keep track of the errors.
 		for _, s := range t.Samples {
-			ref, err = app.Append(ref, lset, s.Timestamp, s.Value)
+			if r.opts.TooFarInFutureTimeWindow != 0 && tooFarInFuture.Before(model.Time(s.Timestamp)) {
```

Comment:
I would suggest creating a new appender and wrapping the existing one when the flag is set to a value > 0. This way we don't need to have a complex if condition here.

Reply:
that's a good idea, implemented.